src/detect-engine-siggroup.c

```diff
@@ -1071,58 +995,35 @@ static int SigGroupHeadTest08(void)
-                               "content:\"test2\"; content:\"test3\"; sid:1;)");
-    if (de_ctx->sig_list == NULL) {
-        result = 0;
-        goto end;
-    }
-    prev_sig = de_ctx->sig_list;
-
-    prev_sig->next = SigInit(de_ctx, "alert tcp any any -> any any "
-                             "(msg:\"SigGroupHead tests\"; content:\"test1\"; "
-                             "content:\"test2\"; content:\"test3\"; sid:2;)");
-    if (prev_sig->next == NULL) {
-        result = 0;
-        goto end;
-    }
-    prev_sig = prev_sig->next;
-
-    prev_sig->next = SigInit(de_ctx, "alert tcp any any -> any any "
-                             "(msg:\"SigGroupHead tests\"; content:\"test1\"; "
-                             "content:\"test2\"; content:\"test3\"; sid:3;)");
-    if (prev_sig->next == NULL) {
-        result = 0;
-        goto end;
-    }
-    prev_sig = prev_sig->next;
-
-    prev_sig->next = SigInit(de_ctx, "alert tcp any any -> any any "
-                             "(msg:\"SigGroupHead tests\"; content:\"test1\"; "
-                             "content:\"test2\"; content:\"test3\"; sid:4;)");
-    if (prev_sig->next == NULL) {
-        result = 0;
-        goto end;
-    }
-    prev_sig = prev_sig->next;
-
-    prev_sig->next = SigInit(de_ctx, "alert tcp any any -> any any "
-                             "(msg:\"SigGroupHead tests\"; content:\"test1\"; "
-                             "content:\"test2\"; content:\"test3\"; sid:5;)");
-    if (prev_sig->next == NULL) {
-        result = 0;
-        goto end;
-    }
-    prev_sig = prev_sig->next;
+    Signature *s = DetectEngineAppendSig(de_ctx, "alert tcp any any -> any any "
+                                                 "(msg:\"SigGroupHead tests\"; content:\"test1\"; "
+                                                 "content:\"test2\"; content:\"test3\"; sid:1;)");
+    FAIL_IF_NULL(s);
+
+    s = s->next = SigInit(de_ctx, "alert tcp any any -> any any "
+                                  "(msg:\"SigGroupHead tests\"; content:\"test1\"; "
+                                  "content:\"test2\"; content:\"test3\"; sid:2;)");
+    FAIL_IF_NULL(s);
+
+    s = s->next = SigInit(de_ctx, "alert tcp any any -> any any "
+                                  "(msg:\"SigGroupHead tests\"; content:\"test1\"; "
+                                  "content:\"test2\"; content:\"test3\"; sid:3;)");
+    FAIL_IF_NULL(s);
+
+    s = s->next = SigInit(de_ctx, "alert tcp any any -> any any "
+                                  "(msg:\"SigGroupHead tests\"; content:\"test1\"; "
+                                  "content:\"test2\"; content:\"test3\"; sid:4;)");
+    FAIL_IF_NULL(s);
+
+    s = s->next = SigInit(de_ctx, "alert tcp any any -> any any "
+                                  "(msg:\"SigGroupHead tests\"; content:\"test1\"; "
+                                  "content:\"test2\"; content:\"test3\"; sid:5;)");
+    FAIL_IF_NULL(s);
 
```

Comment:
I noticed here while trying to use `DetectEngineAppendSig` the tests fail!
I don't know if its the nature of this test itself that doesn't work with `DetectEngineAppendSig` and Array Matching, or is it a significant difference between `SigInit` and `DetectEngineAppendSig` that make one valid than the other in this specific test?

Reply:
It's not clear why `DetectEngineAppendSig` usage is part of this PR as opposed to maintaining the usages of `SigInit`?

Reply:
It is common in the unittests I worked on and ones I've seen to use `DetectEngineAppendSig` it made the tests more compact since it directly reassigns to the latest/appended `Sig` 
Some unittests even have a mixture of both, So I don't really rely on a specific reason for this conversion _other than cleanups_ , yet this test showed a failure upon using `DetectEngineAppendSig` .
So I wonder when to consider leaving `SigInit` and when to convert it?

Reply:
That was a feedback from Victor (see https://github.com/OISF/suricata/pull/5502#discussion_r509016873) that I passed forward, considering, as @ghostinthehive has mentioned, making tests shorter. I'll have a look and try to figure what can be the cause for the test failure in this case!

Reply:
I think you are on the right track! So, debugging, what I got at is that the original test would initialize the signatures following a _queue_ logic, while `DetectEngineAppendSig` appends signatures in a _stack_ logic. That caused the array misalignment. Following that,  I changed the checks to:
```    
FAIL_IF_NOT(sh->init->match_array[0] == de_ctx->sig_list->next->next->next->next);
FAIL_IF_NOT(sh->init->match_array[1] == de_ctx->sig_list->next->next);
FAIL_IF_NOT(sh->init->match_array[2] == de_ctx->sig_list);
```
And the test passed again...

Don't know what is the best approach here, but at least now I guess at least we understand the whys? :)

Reply:
My guess is that either choice is fine. I'd say that if you do decide to keep the `DetectEngineAppendSig`, though, it's indeed better to add the comment that you've mentioned. =]